### DISCUSSION TITLE
How to freeze asset transfer in coinv1

### DISCUSSION THREAD

**Discussion by LawsonGraham** [aptos-labs/aptos-developer-discussions#268] [Questions]
### Discord user ID

lawsg8

### Describe your question in detail.

I want to write a module which can give tokens to users, but have them be non-transferrable (except for by the owner with the capabilities), and then at a later time enable the token to be transferred... I see the freeze cap functions like 

```
   #[legacy_entry_fun]
    /// Freeze a CoinStore to prevent transfers
    public entry fun freeze_coin_store<CoinType>(
        account_addr: address,
        _freeze_cap: &FreezeCapability<CoinType>,
    ) acquires CoinStore {
        let coin_store = borrow_global_mut<CoinStore<CoinType>>(account_addr);
        coin_store.frozen = true;
    }

    #[legacy_entry_fun]
    /// Unfreeze a CoinStore to allow transfers
    public entry fun unfreeze_coin_store<CoinType>(
        account_addr: address,
        _freeze_cap: &FreezeCapability<CoinType>,
    ) acquires CoinStore {
        let coin_store = borrow_global_mut<CoinStore<CoinType>>(account_addr);
        coin_store.frozen = false;
    }
```

are legacy now... is this because FA standard is coming soon? or should these not be used at all? if not, what is best approach here?

### What have you tried or looked at?  Or how can we reproduce the error?

  #[legacy_entry_fun]
    /// Freeze a CoinStore to prevent transfers
    public entry fun freeze_coin_store<CoinType>(
        account_addr: address,
        _freeze_cap: &FreezeCapability<CoinType>,
    ) acquires CoinStore {
        let coin_store = borrow_global_mut<CoinStore<CoinType>>(account_addr);
        coin_store.frozen = true;
    }

    #[legacy_entry_fun]
    /// Unfreeze a CoinStore to allow transfers
    public entry fun unfreeze_coin_store<CoinType>(
        account_addr: address,
        _freeze_cap: &FreezeCapability<CoinType>,
    ) acquires CoinStore {
        let coin_store = borrow_global_mut<CoinStore<CoinType>>(account_addr);
        coin_store.frozen = false;
    }

### Which operating system are you using?

macOS

### Which SDK or tool are you using? (if any)

Aptos CLI

**Comment by kshitijc1506** [comment]
Hey @LawsonGraham, good to hear from you, been a min :) (we interned at Martian together lol). 

AFAIK, those functions are now deprecated and we can't achieve what you're asking for using the Coin standard. Maybe @gregnazario has insight here? 

**edit: They're not deprecated, they're legacy. So you can still use them in the way that David described below :)**

## How to do it (with FA)
[Fungible Asset standard](https://preview.aptos.dev/en/build/smart-contracts/aptos-standards/fungible-asset) is live on mainnet already. You can achieve Fungible Assets that can't be moved like this: 

1. **Use [Dispatchable Fungible Assets](https://x.com/kshitijc_/status/1798569493838569496)**: Live on testnet only for now, but coming to mainnet very soon). 

In your module, create withdraw/deposit hooks, you can invoke custom logic such that you return an error on any given withdraw/deposit. This stops ALL movement of assets between stores. 
**We'll have docs with some guidelines for Dispatchable FA out early next week!** For now, you can see how it's done in this [example](https://github.com/aptos-labs/aptos-core/blob/bbce0f188d7bef08218f1fc1f46bbd3676b5c74b/aptos-move/move-examples/fungible_asset/stablecoin/sources/usdk.move#L190). 

[Here's](https://github.com/aptos-labs/aptos-core/blob/bbce0f188d7bef08218f1fc1f46bbd3676b5c74b/aptos-move/move-examples/fungible_asset/stablecoin/sources/usdk.move#L142) how to register the deposit/withdraw hooks upon asset creation. 

2. Upon creation of the asset, you can **set all stores to be [untransferable](https://github.com/aptos-labs/aptos-core/blob/bbce0f188d7bef08218f1fc1f46bbd3676b5c74b/aptos-move/move-examples/fungible_asset/stablecoin/sources/usdk.move#L116).** 

This way, a malicious actor can't call into `fungible_asset::transfer_store`/`object::transfer` to bypass your smart contract and move stores around. Although, technically, you don't _necessarily_ need this. Users will never be able to move the value stored in the fungible stores anyway. However, probably better to use it anyway to prevent any unexpected behavior (I can think of some scenarios, let me know if you want me to follow up on this thread later). 

## Enabling/Disabling this at the Asset Level at Will
If you want to pause/unpause the disabling of ALL transfers, you can create a `State` resource within the FA metadata object with a bool (`transfers_paused`) field that toggles this. This isn't exactly the same thing, but here's [something](https://github.com/aptos-labs/aptos-core/blob/bbce0f188d7bef08218f1fc1f46bbd3676b5c74b/aptos-move/move-examples/fungible_asset/stablecoin/sources/usdk.move#L45) for reference. 

Then, create 2 functions:

1. [`is_state_paused()`](https://github.com/aptos-labs/aptos-core/blob/bbce0f188d7bef08218f1fc1f46bbd3676b5c74b/aptos-move/move-examples/fungible_asset/stablecoin/sources/usdk.move#L324) and add this logic to your withdraw/deposit hooks. Verify if transfers are not paused before allowing a deposit. 
2. [`un/pause_state()`](https://github.com/aptos-labs/aptos-core/blob/bbce0f188d7bef08218f1fc1f46bbd3676b5c74b/aptos-move/move-examples/fungible_asset/stablecoin/sources/usdk.move#L265), which retrieves the metadata object and the State resource, and toggles the bool field. 

When reading through the example, pls note that the use case the usdk.move module is trying to achieve is different. It doesn't prevent the movement of all assets, but rather just aims to denylist specific users. So the functionality only extends to denylisted users. 

## Using Digital Asset Standard Instead
This is another option, you can create semi/non-fungible Soulbound tokens with the new DA standard. Here are the [docs](https://preview.aptos.dev/en/build/smart-contracts/aptos-standards/digital-asset), feel free to ask me any question on implementation details here :)

**Comment by davidiw** [comment]
So for coins and fungible assets this is possible in both the same ways assuming the following properties:
1. Unlocked tokens can go anywhere
2. Only an admin or a limited function can create or move assets

So the steps:
1. The minter needs access to the transfer ref/ freeze cap
2. When minting tokens, they freeze the store where they are being deposited
3. For movement, they can either eventually become unlocked and moved freely or can only be moved through a function that has access to the freeze cap / transfer ref

**Reply by LawsonGraham** [reply]
got it - this makes sense. thanks!
